templates/README.md

```diff
@@ -0,0 +1,254 @@
+
+    # This set of annotations is meant as a hint only, it is
+    # not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded here informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+    defaults.template.cnv.io/disk: default-disk
+    defaults.template.cnv.io/volume: default-volume
+    defaults.template.cnv.io/network: default-network
+
+    # Replacements useful for finishing the transformation
+    # This is an extension we could use for specifying common
+    # fields for all disks/nics. Requires extra implementation
+    # that will do the find and replace using a syntax documented
+    # here: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+    # The jsonpath root is the objects: element of the template
+    # XXX There is currently an issue with annotations not supporting
+    #     list values
+    template.cnv.io/replace:
+    - path: "/VM/disks/*/type"
+      value: virtio
+
+    # Another extension useful for CNV aware tooling that allows
+    # expressing additional validation rules for this template
+    # The jsonpath root is the objects: element of the template
+    # XXX There is currently an issue with annotations not supporting
+    #     list values
+    template.cnv.io/validate:
+    - path: "/VM/spec/domain/guest/memory"
+      min: 3
+      max: 4
+
+    # Another extension for hinting at which elements should be
+    # considered editable. The content is a line separated
+    # list of jsonpath selectors.
+    # The jsonpath root is the objects: element of the template
+    template.cnv.io/editable: |
+      /objects[0].spec.template.spec.domain.cpu.cores
+      /objects[0].spec.template.spec.domain.resources.requests.memory
+      /objects[0].spec.template.spec.domain.devices.disks
+      /objects[0].spec.template.spec.volumes
+      /objects[0].spec.template.spec.networks
+
+  labels:
+    # The UI can show all possible template.cnv.io/* values in a nice way
+    # and let the user filter down the available templates to the one
+    # the user actually wants:
+    # A single selected template only means no conflicts and no smart
+    # merging code. This has to be done using labels to allow efficient
+    # searching.
+    # The format has the following meaning:
+    # {os,flavor,size}.template.cnv.io/{value}: true (or false for exclusion)
+    # OS names should match the libosinfo identifiers
+    # flavors are tiny, medium, large, etc.
+    # workloads are generic, high-performance, io-intensive,
+    #               oracle-db, sap-hana...
+    os.template.cnv.io/windows2k12r2: "true"
+    os.template.cnv.io/windows8: "true"
+    os.template.cnv.io/windows7: "true"
+    workload.template.cnv.io/minimal: "true"
+    workload.template.cnv.io/io-intensive: "true"
+    # flavor.template.cnv.io/* not specified means all
```

Comment:
I dont think we need it? I think it would be easier if the UI would simply filter whatever is behind the '/'. If we will have this notion of "all" and "all but" it means we need to have an independent database of the values which we would need to maintain etc

Reply:
Or just prepare the in memory list of allowed fields by loading all templates first.

Reply:
@MarSik why instead not:

```
os: windows7
flavor: large
workload: io-intensive
```

I'd rather have three templates which are slightly different i.e. to reflect large, small, medium, in favor of a simple tempalte and nice labels.

Reply:
Because you will not have three files, but thirty. And rhel7.0 is going to be identical to rhel7.1, rhel7.2, ... , rhel7.5.

Reply:
Yes, yes, yes ...

Reply:
Along those lines: We could see that we get "linux families" or "tribes" which the reflect
"ubuntu", "rhel", "fedora"
which canbe a generic base.

Reply:
Can we actually have some inheritance of workload/profiles?
I mean, we can set, for instance:
```
os: fedora21
flavor: larger
workload: io-intensive
```

And for the next one's we could have something like:
 ```
os: fedora22
derives-from: fedora21
```

Those could give us, for free, the flavor workload we want.
This is basically what we do on osinfo-db regarding the supported devices.

Of course, going for families is also an options.

Reply:
No inheritance in the shipped templates. That would mean all the runtime tooling needs to load the whole chain. Instead you can add multiple OSes to the same template. Also the limitation here is in what Openshift template mechanism supports. The templates must work in CLI and CI automation environments with what we currently have. 

However, the plan is to compile the final templates from snippets using some kind of source code level templating and compile them into the final form before release.